KEP 993: Two Stage Admission Process (#994)
kubernetes-sigs/kueue: keps/993-two-phase-admission/README.md

#### Story 1

I want to use the [Cluster Autoscaler ProvisioningRequest API](https://github.com/kubernetes/autoscaler/pull/5848) 
to ensure that the resources can be provided in full before starting the workload.  

#### Story 2

I want to base workload admission on budget. I have the actual consumption metrics
in my Prometheus/Datadog/Stackdriver/whatever monitoring system. And weekly
budget in a CRD, per namespace. I’m happy to modify some small existing sample
controller, but I don’t want to fork the whole Kueue to adjust some minor details.
 
### Notes/Constraints/Caveats (Optional)

Each of the go/no-go decisions is best-effort and can be reversed at any time. 

### Risks and Mitigations

## Design Details

We will extend ClusterQueueSpec definition by adding a reference to
AdmissionChecks that the queue needs to perform before admitting a
workload. The design of AdmissionChecks is similar to IngressClass
- it allows passing additional parameters, specific to particular checks
that need to be performed. 

```
// ClusterQueueSpec defines the desired state of ClusterQueue
type ClusterQueueSpec struct {
[...]
	// List of AdmissionCheck names that needs to be passed

> 	// to admit a workload.  

Author: Done.

> 	AdmissionChecks []string `json:"addotomalChecks"`

Author: Done

}

// Cluster-scoped top-level object defining a check that 
// can be referenced in ClusterQueue
type AdmissionCheck struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`
	Spec AdmissionCheckSpec `json:"spec,omitempty"`
}

const (
	Retry CheckFailureState = “Retry”
	Reject CheckFailureState = “Reject”
)

type AdmissionCheckSpec struct {
// Name of the controller which will actually perform
// the checks. Cannot be empty.
	ControllerName string `json:"controllerName"`

	// How long to keep the workload suspended 

> 	// after a failed check.

Author: Yes. Comment expanded.

	RetryDelayMinutes *int64. `json:"retryDelayMinutes,omitempty"`

	// A reference to the additional parameters for the check. 
	Parameters *AdmissionCheckParametersReference `json:"parameters,omitempty"`
}

// Points to where the parameters for the checks are. 
// As clusterqueue are in cluster scope - this should be
// a dedicated CRD specific for the Controller.
type AdmissionCheckParametersReference struct {
	// ApiGroup is the group for the resource being referenced. 
	APIGroup string `json:"apiGroup"`

	// Kind is the type of the resource being referenced.
	Kind string `json:"kind""`

	// Name is the name of the resource being referenced.
	Name string `json:"name"`
}
```

For every workload that is put to a ClusterQueue, Kueue will perform the very 
same checks that it does today, before admitting a workload. However, once the basic checks pass AND there are some AdmissionChecks configured, AND the workload is not in on-hold retry state from some check, it will:

1. Fill the Admission field in workload, with the desired flavor assignment. 

> 2. Not do any preemptions yet.

Reviewer: This one is tricky. Let's take the scenario of ProvisioningRequest.

If a Workload requires preemptions to fit and we don't do preemptions, the ProvisioningRequest would likely trigger a scale up or fail because it reached the scale up limits (or reached stockout).

OTOH, if we do the preemptions, and the ProvisioningRequest cannot be satisfied for whatever reason, we would have done preemptions for nothing (or too early). Not sure if this should be a configuration parameter for the check.

Another open question: if we defer preemptions, should any more workloads be admitted in the cohort while waiting for the admissionChecks? In general I think we should not. Or perhaps it should be per flavor/resource that needs preemption or is borrowing, like in #1036.

Reviewer 2: +1 for this being tricky, and also not very easy to do with the current preemption mechanism.
On the overall two steps admission, I propose, as a first stage to not change the way the preemption is done.

In the future, maybe we can add a built in `AdmissionCheckPreempt`, the controller of which should be in charge of issuing evictions and wait for the resources to become available.

Author: We cannot start with preemptions - if we preempt we will create a "hole" in in the quota and the preempted workload will try to get in again immediately. Thus preemption can only work with simultaneous admission or other form of the resource reservation. Thus if we preeempt, we should account the precheckedWorkloads in the quota, that basically means that we cannot have best effort queues as we have right now, only "moderate effort" ;).

Author: In general, we probably need 3 modes of operations:

* Preempt after checks pass - for checks that are not fully correlated with quota, and can take a longer time to give a green light.
* Preempt before checks are started - for checks that are essentially yet another quota check.
* Mix of the two based on the status provided by the check. The check may say:
   * Preempt now, as i'm hitting quota issues and preemption may help
   * Preempt after I give a green light, no quota issues observed at this moment and the admission will take a longer time, so there is no need to stop the running workloads yet. 

I will update the document to reflect that.

3. Set conditions on the workload:
* “AdmissionPrecheck” to true

> * “AdmissionCheck|<checkName>” to unknown for each of the add

Reviewer: Validation here might be tricky:
`checkName` has to be a valid CRD name. Whereas the condition type has to be a valid qualified name. Not sure if one is a subset of the other.
But overall I think it's worth just having a dedicated field for the admissionCheck statuses.

That also avoids the external controllers to concatenate an "AdmissionCheck" prefix into their condition name.

Author: Ok, added.

> * “AdmissionCheck|<checkName>” to unknown for each of the add

Reviewer 2: Maybe it makes sense to have list checks populated and maintained throughout the entire life of the workload, this way the check controller can choose if it wants to wait for the kueue admission to be set before it updates it's specific status.

Author: OK. Updated.

Kueue will only pass as many pods past AdmissionPrecheck as there would
fit in the quota together, possibly limited by a Kueue configuration option.
That would violate a bit BestEffort logic of queues - if there is 1000 tasks
in a queue that doesn’t pass quota and 1001st task passes, the best effort
queue would let it in. BestEffort queue with 1000 tasks that don’t pass

> AdditioanlChecks would not let 1001st task in (it will not reach the

Reviewer: ```suggestion
admissionChecks would not let 1001st task in (it will not reach the
```

Author: Done.

checks). Without this limitation, a large number of tasks could be switched

> back and forth from Precheck to suspended state. 

Reviewer: When we have determined that checks for a workload don't pass (and the retry period is over), we should probably put the workload in the `inadmissibleWorkloads` set.
This way, BestEffort would mostly work the same way as before, except that it's, potentially, slower.

Author: Ack.

budget in a CRD, per namespace. I’m happy to modify some small existing sample
controller, but I don’t want to fork the whole Kueue to adjust some minor details.
 
### Notes/Constraints/Caveats (Optional)

Each of the go/no-go decisions is best-effort and can be reversed at any time. 

### Risks and Mitigations

## Design Details

We will extend ClusterQueueSpec definition by adding a reference to
AdmissionChecks that the queue needs to perform before admitting a
workload. The design of AdmissionChecks is similar to IngressClass
- it allows passing additional parameters, specific to particular checks
that need to be performed. 

```
// ClusterQueueSpec defines the desired state of ClusterQueue
type ClusterQueueSpec struct {
[...]
	// List of AdmissionCheck names that needs to be passed
	// to admit a workload. The order of the checks doesn't matter,
	// all of them will be started at the same time.
	AdmissionChecks []string `json:"admissionChecks"`
}

// WorkloadStatus defines the observed state of Workload
type WorkloadStatus struct {
[...]
	// Status of admission checks, if there are any specified 
	// for the ClusterQueue in which the workload is queued.

> 	AdmissionChecksStatuses []metav1.Condition `json:"admissionChecksStatuses,omitempty"`

Author: Ok, changed.

}

// Cluster-scoped top-level object defining a check that 
// can be referenced in ClusterQueue

> type AdmissionCheck struct {

Author: v1beta1

> type AdmissionCheck struct {

Reviewer: If we go beta, we can still change the API, but we need to keep support for the old API for some releases.

I'm ok with alpha or beta.

	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`
	Spec AdmissionCheckSpec `json:"spec,omitempty"`
}

const (
	Retry metav1.ConditionStatus = “Retry”

> 	Reject metav1.ConditionStatus = “Reject”

Author: Done

	QuotaExhausted metav1.ConditionStatus = "QuotaExhausted"

> )

Reviewer: nit, move before `AdmissionChecksStatuses`, so that it's more clear that it's about them (IIUC).

But in general, explain what they mean.

Author: Explanations added.

> )

Author: Yes, they are to be used in addition of true, false and unknown.

> )

Author: Actually, after @trasc comment, we will put them in Reasons. Updated the doc.

> type AdmissionCheckSpec struct {

Author: Added.

// Name of the controller which will actually perform
// the checks. Cannot be empty.

> 	ControllerName string `json:"controllerName"`

Reviewer 3: Is this `controllerName` similar to the `controller` field in the ingressClass?

https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-class

> 	ControllerName string `json:"controllerName"`

Author: Yes, similar to the one in ingress-class.

> 	ControllerName string `json:"controllerName"`

Author: Added some explanation.

	// How long to keep the workload suspended 

> 	// after a failed check (after it transitioned to False).

Reviewer: When this time is over, would we transition back to Unknown? That could be a good signal to requeue.

Author: Yes, added.

> 	// after a failed check (after it transitioned to False).

Author: Let's say 15 min. Added.

> 	// after a failed check (after it transitioned to False).

Author: It means that the workload goes back to Suspended state and will not be retried for the given time period. Other workloads may proceed if the queue type is other than strict Fifo.

> 	RetryDelayMinutes *int64. `json:"retryDelayMinutes,omitempty"`

Author: Done.

	// A reference to the additional parameters for the check. 
	Parameters *AdmissionCheckParametersReference `json:"parameters,omitempty"`

	// If set to true, the workload will preempt other and block
	// capacity before the check is passed.

> 	BookCapacity bool `json:"bookCapacity,omitempty"`

Reviewer: ```suggestion
	BookQuota bool `json:"bookQuota,omitempty"`
```

Reviewer: But could it be an enum? https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types

Also, what happens if different checks have different settings?

> 	BookCapacity bool `json:"bookCapacity,omitempty"`

Author: The most restrictive option wins.

}

// Points to where the parameters for the checks are. 
// As clusterqueue are in cluster scope - this should be
// a dedicated CRD specific for the Controller.
type AdmissionCheckParametersReference struct {
	// ApiGroup is the group for the resource being referenced. 
	APIGroup string `json:"apiGroup"`

	// Kind is the type of the resource being referenced.

> 	Kind string `json:"kind""`

Author: Done.

	// Name is the name of the resource being referenced.
	Name string `json:"name"`
}
```

For every workload that is put to a ClusterQueue that has admissionChecks
configured Kueue will add admissionCheckStatuses:

> * “AdmissionPrecheck” to false

Reviewer: ```suggestion
* “AdmissionPrecheck” in `status.conditions` to false
```

Author: Done.

> * “AdmissionPrecheck” to false

Author: Done.

> * “<checkName>” to unknown for each of the AdmissionChecks

Reviewer: ```suggestion
* “<checkName>” in `status.admissionCheckStatuses` to unknown for each of the AdmissionChecks
```

Kueue will perform the very same checks that it does today, 

> before admitting a workload. However, oncethe basic checks 

Reviewer: ```suggestion
before admitting a workload. However, once the basic checks 
```

Author: Done.

Kueue will only pass as many pods past AdmissionPrecheck as there would
fit in the quota together, possibly limited by a Kueue configuration option.
That would violate a bit BestEffort logic of queues - if there is 1000 tasks
in a queue that doesn’t pass quota and 1001st task passes, the best effort
queue would let it in. BestEffort queue with 1000 tasks that don’t pass
AdmissionChecks would not let 1001st task in (it will not reach the
checks). Without this limitation, a large number of tasks could be switched
back and forth from Precheck to suspended state. 

Then it will wait until all additional check conditions are satisfied (set
to true). Once it happens, the Kueue will recheck that Admission settings
are still valid (check quota/preemptions/reclamation) and admit the
workload (doing preemptions if needed). Kueue will also recheck all of the
prechecked workloads on every other workload admission, to make sure
that their flavor admission is still possible. 

If any of these checks fails (either on the final check or when admission
other workloads), the workload gets back to the suspended state, and gets
their AdmissionChecks and AdmissionPrecheck cleared.

> If any check is switched to Reject or Retry, the workload is either

Reviewer: Completely rejected means that we will mark the Job as failed?

Author: Yes, and drop from the queue (so that other workloads can proceed).

completely rejected or switched back to suspend state (with exponential

> backoff retry delay). QuotaExhausted means that the quota is reached

Reviewer: this might be confusing, as quota is a term in Kueue. IIUC, here you are referring to an autoscaling quota.

Maybe we can be more generic and say: `BookQuotaRequired`, similar to the `BookQuota` field in the AdmissionCheckSpec

Author: Ok, changed.

and Kueue should try to preempt some workloads, before the check can
succeed.

The controller implementing a particular check should:
 
* Watch all AdmissionCheck objects to know which one should be
handled by it. 
* Watch all controller specific parameter objects, potentially referenced
from AdmissionCheck.
* Watch all workloads and process those that have AdmissionCheck for this
particular controller and are past AdmissionPrecheck.
* After approving the workload, keep an eye on the check if it starts failing,
fail the check and cause the workload to move back to the suspended state.

### Test Plan

[ x ] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

#### Unit Tests

#### Integration tests

> The tests should cover:

Reviewer: Also: More than one addmisionCheck

Author: Added.

	// to admit a workload. The order of the checks doesn't matter,
	// all of them will be started at the same time.
	AdmissionChecks []string `json:"admissionChecks"`
}

// WorkloadStatus defines the observed state of Workload
type WorkloadStatus struct {
[...]
	// Status of admission checks, if there are any specified 
	// for the ClusterQueue in which the workload is queued.
	AdmissionChecksConditions []metav1.Condition `json:"admissionChecksConditions,omitempty"`
}

// Cluster-scoped top-level object defining a check that 
// can be referenced in ClusterQueue
type AdmissionCheck struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`
	Spec AdmissionCheckSpec `json:"spec,omitempty"`
}

const (
	// The ckech cannot pass at this moment, back off (possibly 
	// allowing other to try, unblock quota) and retry.
	Retry metav1.ConditionStatus = "Retry"
	// The check will not pass in the near future. It is not worth
	// to retry.
	Reject metav1.ConditionStatus = "Reject"
	// The check might pass if there was fewer workloads. Proceed
	// with any pending workload preemption.
	QuotaExhausted metav1.ConditionStatus = "QuotaExhausted"

> )

Reviewer: @mwielgus we either go with Reasons or create our own struct, instead of using Conditions. I'm fine with either. WDYT?

> )

Author: OK.

> )

Author: Updated.

> )

Author: We can stay with Conditions.

	// before the check pass. The default is NoBooking.
	// If different modes are used for different AdmissionChecks,
	// the most restrictive one is used.
	QuotaBooking *QuotaBookingMode `json:"quotaBooking,omitempty"`
}

const (
	// No booking, unless a check explicitely asks for it reporting quota exhaused.
	NoBooking QuotaBookingMode = "NoBooking"
	
	// Book quota (and preempt) immediately after the basic quota checks 
	// are done, without waiting for the additional checks to cpmplete.
	BookAfterQuotaCheck QuotaBookingMode = "BookAfterQuotaCheck"
)

// Points to where the parameters for the checks are. 
// As clusterqueue are in cluster scope - this should be
// a dedicated CRD specific for the Controller.
type AdmissionCheckParametersReference struct {
	// ApiGroup is the group for the resource being referenced. 
	APIGroup string `json:"apiGroup"`

	// Kind is the type of the resource being referenced.
	Kind string `json:"kind"`

	// Name is the name of the resource being referenced.
	Name string `json:"name"`
}
```

For every workload that is put to a ClusterQueue that has AdmissionChecks
configured Kueue will add:

> * “AdmissionPrecheck” set false to Conditions.

Author: I would prefer to stay with Precheck - in near future we may let a bit more workloads past the Precheck than the quota would allow (and assume that the Check will effectively throttle/queue/apply quota to them).

> * “AdmissionPrecheck” set false to Conditions.

Reviewer: +1 on Marcin. The quota might be violated if we defer preemption.

> * “AdmissionPrecheck” set false to Conditions.

Reviewer: Actually, to follow https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties,
It should be AdmissionPrechecked

> * “AdmissionPrecheck” set false to Conditions.

Reviewer: From @trasc in https://github.com/kubernetes-sigs/kueue/pull/1075#discussion_r1303223834

> There are two reasons why I don't think it's a good idea:

> 1. This can imply that is done before the checks.

This is true. We want a Condition that, when true, `admissionChecks` can start.

> 2. I think, the Precheck, as described in the KEP is not blocking the quota

Correct, and that's why naming it QuotaReserved doesn't seem accurate. Unless you are suggesting having 3 conditions?
Can you elaborate?

> * “AdmissionPrecheck” set false to Conditions.

Reviewer 2: > This is true. We want a Condition that, when true, `admissionChecks` can start.

There is no reason to block the admissionCheck until the Quota is reserved/checked, a controller can choose to do it only after, but i don't see any value forcing this.

> Correct, and that's why naming it QuotaReserved doesn't seem accurate. Unless you are suggesting having 3 conditions?
Can you elaborate?

In #1045 having the condition (which is still called Admitted there, #1075 only renames it) means that the quota is blocked for the workload in question. 

 Now, if there is a usecase in which a check controller needs to know that at some point in the past the workload wold have had the option to block the quota, what it could do is. 

1. Keep it's condition in a non False state (preferably Unknown)
2. The scheduler will reserve the quota at some point. 
3. The controller can switch it's condition to false. (this will cancel the quota reservation)
4. Perform it's check. And set its condition to True.
5. The workload will go into the scheduling again.

(if the controller has a quick way of doing the check in some cases, the reservation cancelling and the second scheduling can be skipped)

> * “AdmissionPrecheck” set false to Conditions.

Reviewer: Let's recap.
In implementation terms, what we need is a condition that indicates that Kueue has executed the scheduling cycle and, if necessary, found candidates for preemption.

Preemption might or might not happen at this point, depending on what is set in `quotaBooking` . If this setting says `NoBooking`, meaning that we shouldn't do preemption, then it doesn't make sense to call the condition `QuotaReserved`.

`PreCheck` is a bit more neutral, although a bit vague. Let me think about it more.

> * “AdmissionPrecheck” set false to Conditions.

Reviewer: How about `NominatedForAdmission`?
- It goes along with the fact that we fill up the `admissionField`.
- It doesn't give any indication that `admissionChecks` run before or after.
- It doesn't give any indication about whether preemption happened or quota is reserved.

> * “AdmissionPrecheck” set false to Conditions.

Reviewer 2: But it has no indication that the quota is blocked.

	Reject string = "Reject"
	// The check might pass if there was fewer workloads. Proceed
	// with any pending workload preemption. Should be set 
	// when condition status is still "Unknown" 
	BookQuotaRequired status = "BookQuotaRequired"
)

// AdmissionCheckSpec defines the desired state of AdmissionCheck.
type AdmissionCheckSpec struct {
	// Name of the controller which will actually perform
	// the checks. This is the name with which controller indentifies with,
	// not a K8S pod or deployment name. Cannot be empty. 
	ControllerName string `json:"controllerName"`

	// How long to keep the workload suspended 
	// after a failed check (after it transitioned to False).
	// After that the check state goes to "Unknown".
	// The default is 15 min. 
	RetryDelayMinutes *int64 `json:"retryDelayMinutes,omitempty"`

	// A reference to the additional parameters for the check. 
	Parameters *AdmissionCheckParametersReference `json:"parameters,omitempty"`

	// How should the workload book quota (and possibly preempt others) 
	// before the check pass. The default is NoBooking.
	// If different modes are used for different AdmissionChecks,

> 	// the most restrictive one is used.

Author: Added.

	QuotaBooking *QuotaBookingMode `json:"quotaBooking,omitempty"`
}

const (
	// No booking, unless a check explicitely asks for it reporting quota exhausted.
	NoBooking QuotaBookingMode = "NoBooking"
	
	// Book quota (and preempt) immediately after the basic quota checks 
	// are done, without waiting for the additional checks to cpmplete.

> 	BookAfterQuotaCheck QuotaBookingMode = "BookAfterQuotaCheck"

Author: That is potentially misleading. Some "NoBooking" checks may be executed after the booking has actually been done.

> 	BookAfterQuotaCheck QuotaBookingMode = "BookAfterQuotaCheck"

Reviewer 2: > +1 on Marcin. The quota might be violated if we defer preemption.

There should be a new check (managed the built in controller discussed offline) which, regardless to when the eviction is triggerd will be set to ready only after the necessary eviction process ended.

)

// Points to where the parameters for the checks are. 
// As clusterqueue are in cluster scope - this should be
// a dedicated CRD specific for the Controller.
type AdmissionCheckParametersReference struct {
	// ApiGroup is the group for the resource being referenced. 
	APIGroup string `json:"apiGroup"`

	// Kind is the type of the resource being referenced.
	Kind string `json:"kind"`

	// Name is the name of the resource being referenced.
	Name string `json:"name"`
}
```

For every workload that is put to a ClusterQueue that has AdmissionChecks
configured Kueue will add:

* “AdmissionPrecheck” set false to Conditions.
* “<checkName>” set to "Unknown" for each of the AdmissionChecks to AdmissionCheckConditions.

Kueue will perform the very same checks that it does today, 
before admitting a workload. However, once the basic checks 
pass AND there are some AdmissionChecks configured, AND the
workload is not in on-hold retry state from some check, it will:

> 1. Fill the Admission field in workload, with the desired flavor assignment. 

Reviewer: This should happen before, along with setting AdmissionPrecheck, otherwise we could lose the assignment on a controller restart.

However, at this point, we should set the Admitted condition to true.

Reviewer: Nvm, I got confused with the paragraphs.

The current proposal makes sense.
